agent/asm/asm.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//    http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package asm
+
+import (
+	"encoding/json"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/secretsmanager"
+	"github.com/aws/aws-sdk-go/service/secretsmanager/secretsmanageriface"
+	docker "github.com/fsouza/go-dockerclient"
+	"github.com/pkg/errors"
+)
+
+// AuthDataValue is the schema for
+// the SecretStringValue returned by ASM
+type AuthDataValue struct {
+	Username *string
+	Password *string
+}
+
+// GetDockerAuthFromASM makes the api call to the AWS Secrets Manager service to
+// retrieve the docker auth data
+func GetDockerAuthFromASM(secretID string, client secretsmanageriface.SecretsManagerAPI) (docker.AuthConfiguration, error) {
+	in := &secretsmanager.GetSecretValueInput{
+		SecretId: aws.String(secretID),
+	}
+
+	out, err := client.GetSecretValue(in)
+	if err != nil {
+		return docker.AuthConfiguration{}, errors.Wrapf(err,
+			"asm fetching secret from the service for %s", secretID)
+	}
+
+	return extractASMValue(out)
+}
+
+func extractASMValue(out *secretsmanager.GetSecretValueOutput) (docker.AuthConfiguration, error) {
+	if out == nil {
+		return docker.AuthConfiguration{}, errors.Errorf(
+			"asm fetching authorization data: empty response")
+	}
+
+	secretValue := aws.StringValue(out.SecretString)
+	if secretValue == "" {
+		return docker.AuthConfiguration{}, errors.Errorf(
+			"asm fetching authorization data: empty secrets value")
+	}
+
+	authDataValue := AuthDataValue{}
+	err := json.Unmarshal([]byte(secretValue), &authDataValue)
+	if err != nil {
+		// could  not unmarshal, incorrect secret value schema
+		return docker.AuthConfiguration{}, errors.Wrapf(err,
```

Comment:
is it always safe to print this err? Does the json unmarshal error sometimes return parts of the original string? I'd lean toward not printing it.

Reply:
> I'd lean toward not printing it.

why exactly? this is an error that would be passed back up to CP as reason for why the the auth data wasn't acquired.

Reply:
It might print some of the secret text, which we don't want.

Reply:
hm i can't seem to find anything that confirms that errors will *not* contain partials. probably best to leave it out then?